src/UserSpaceInstrumentation/AllocateInTracee.h

```diff
@@ -9,30 +9,92 @@
+  // We move around MemoryInTracee exclusively in unique_ptr so we don't need it to be copyable or
+  // movable. The only way to create a MemoryInTracee is to use the factory function below which
+  // uses the private constructor.
+  MemoryInTracee() = delete;
+  MemoryInTracee(const MemoryInTracee&) = delete;
+  MemoryInTracee(MemoryInTracee&&) = delete;
+  MemoryInTracee& operator=(const MemoryInTracee&) = delete;
+  MemoryInTracee& operator=(MemoryInTracee&&) = delete;
+
+  // Allocate `size` bytes of memory in the tracee's address space using mmap. The memory will have
+  // write permissions. The memory allocated will start at `address`. `address` needs to be aligned
+  // to page boundaries. If the memory mapping can not be placed at `address` an error is returned.
+  // If `address` is zero the placement of memory will be arbitrary (compare the documentation of
+  // mmap: https://man7.org/linux/man-pages/man2/mmap.2.html). Assumes we are already attached to
+  // the tracee `pid` using `AttachAndStopProcess`.
+  [[nodiscard]] static ErrorMessageOr<std::unique_ptr<MemoryInTracee>> Create(pid_t pid,
+                                                                              uint64_t address,
+                                                                              uint64_t size);
+
+  // Free address range previously allocated with AllocateInTracee using munmap.
+  // Assumes we are already attached to the tracee using `AttachAndStopProcess`.
+  [[nodiscard]] ErrorMessageOr<void> Free();
+
+  [[nodiscard]] pid_t GetPid() const { return pid_; }
+  [[nodiscard]] uint64_t GetAddress() const { return address_; }
+  [[nodiscard]] uint64_t GetSize() const { return size_; }
+  // Returns the protection state of the memory. The memory gets allocated as writeable and can be
+  // made executable (and writable again) with the methods below.
+  enum class State { kWritable, kExecutable };
+  [[nodiscard]] State GetState() const { return state_; }
+
+  // Sets the read and execute permission for the memory. Removes the write permission. Assumes we
+  // are already attached to the tracee using `AttachAndStopProcess`.
+  [[nodiscard]] ErrorMessageOr<void> EnsureMemoryExecutable();
+  // Set the write permission for the memory. Removes the read and execute permissions. Assumes we
+  // are already attached to the tracee using `AttachAndStopProcess`.
+  [[nodiscard]] ErrorMessageOr<void> EnsureMemoryWritable();
+
+ protected:
+  MemoryInTracee(pid_t pid, uint64_t address, uint64_t size, State state)
+      : pid_(pid), address_(address), size_(size), state_(state) {}
+
+  pid_t pid_ = -1;
+  uint64_t address_ = 0;
+  uint64_t size_ = 0;
+  State state_ = State::kWritable;
+};
+
+// Same as `MemoryInTracee` above but deallocates memory in the destructor. Note that we still need
+// to be attached (or attached again) to the tracee when `AutomaticMemoryInTracee` goes out of
+// scope.
+class AutomaticMemoryInTracee : public MemoryInTracee {
```

Comment:
personal preference. I am not a big fun on inheritance in OOP, it seems like this could be a flag.

Reply:
I think it would be clearer if there was a base abstract type `MemoryInTracee` and then two subclasses `AutomaticMemoryInTracee` and `LeakableMemoryInTracee` (or whatever).
In the current state `AutomaticMemoryInTracee`'s dtor is replacing the behavior of the superclass's dtor and it's better to avoid these situations.

Reply:
> seems like this could be a flag.

I thought about implementing this with a with enum{leaky,automatic} as a template parameter and then possibly typedef'ing the classes. But there really isn't any difference.

>In the current state AutomaticMemoryInTracee's dtor is replacing the behavior of the superclass's dtor and it's better to avoid these situations.

Is there a reason I am missing? Why should I avoid that? After all this is exactly where I want the both classes to differ.

Reply:
> Is there a reason I am missing? Why should I avoid that?

Overriding a non-abstract method (this case is special because it's a destructor but it doesn't really matter) often makes the concept of is-a relationship that public inheritance is expressing a bit wonky.
In this case we know that `MemoryInTracee` is used as a handle to an area of memory that should be leaked. The fact that we know this class will be used to leak the memory even if not "advertised" as such is key here. Since `AutomaticMemoryInTracee` is used as a handle to an area of memory to be freed up, this creates a contradiction as the public inheritance is saying "`AutomaticMemoryInTracee` is a (to-be-leaked)`MemoryInTracee`".
If instead you have `MemoryInTracee` just as an interface for a handle to an area in memory, and then two subclasses with the two different semantics, the relationships between the classes are cleaner.
You would also simplify the interface a bit as you don't need to expose `Free` publicly anymore, by saying that either a type is used to leak the memory, or the type is used to free it up on destruction (RAII-like).
Let's not hesitate to discuss offline if this is getting difficult to discuss here.
Maybe @dimitry- also has an opinion in case I'm looking at this from the wrong point of view.

Reply:
Okay, I guess this really just a matter of looking at it: MemoryInTracee is a class representing memory, AutomaticMemoryInTracee is a MemoryInTracee with the additional trait/functionality/whatever that it deallocates on destruction.

Reply:
Discussed offline and we agreed to disagree :D